cpp/inline/games/stochastic_nim/PerfectPlayer.inl

```diff
@@ -0,0 +1,66 @@
+#include <games/stochastic_nim/PerfectPlayer.hpp>
+
+namespace stochastic_nim {
+
+inline PerfectPlayer::ActionResponse PerfectPlayer::get_action_response(const State& state,
+                                                                 const ActionMask& valid_actions) {
+  if (state.current_mode == kChanceMode) {
+    throw std::invalid_argument("PerfectPlayer does not support chance mode");
+  }
+
+  if (state.stones_left > strategy_->get_params().starting_stones) {
+    throw std::invalid_argument("PerfectPlayer does not support more stones than starting stones");
+  }
+
+  return strategy_->get_optimal_action()[state.stones_left] - 1;
```

Comment:
This is a bit of an abstraction leak. The detail that `PerfectStrategy` stores its action as 1-indexed rather than 0-indexed doesn't deserve to extend beyond the `PerfectStrategy` abstraction boundary.

Better is,

```
return strategy_->get_optimal_action(state.stones_left);
```